src/Language/DifferentialDatalog/Compile.hs

```diff
@@ -3316,9 +3316,15 @@ mkExpr' d ctx e@EBinOp{..} = (v', EVal)
     t2 = exprType' d (CtxBinOpR e' ctx) (E $ sel3 exprRight)
     v = case exprBOp of
              Concat | t == tString
-                    -> case sel3 exprRight of
-                            EString _ s -> "::ddlog_rt::string_append_str(" <> e1 <> ", r###\"" <> pp s <> "\"###)"
-                            _           -> "::ddlog_rt::string_append(" <> e1 <> "," <+> ref exprRight <> ")"
+                    -> -- The parser scans expressions like '"${x}"' as '"" ++ x.to_string()'.
+                       -- The empty string operand helps type inference figure
+                       -- out that this is string concatenation, so we cannot
+                       -- eliminate it in the parser.  Drop it here instead.
+                       if E (sel3 exprLeft) == eString ""
+                       then val exprRight
```

Comment:
Technically you can drop empty strings in the right too.
There should be a strength reduction pass for this kind of stuff.
I guess rust is not smart enough to do it.